engine/runtime-integration-tests/src/test/java/org/enso/interpreter/test/PrivateAccessTest.java

```diff
@@ -0,0 +1,113 @@
+import static org.hamcrest.CoreMatchers.containsString;
+import static org.hamcrest.CoreMatchers.is;
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.junit.Assert.fail;
+
+import java.io.IOException;
+import java.nio.file.Files;
+import java.nio.file.Path;
+import org.enso.polyglot.PolyglotContext;
+import org.enso.polyglot.RuntimeOptions;
+import org.graalvm.polyglot.PolyglotException;
+import org.junit.Rule;
+import org.junit.Test;
+import org.junit.rules.TemporaryFolder;
+
+public class PrivateAccessTest extends TestBase {
+  @Rule public TemporaryFolder tempFolder = new TemporaryFolder();
+
+  /**
+   * Private constructor defined in a type inside unknown module that has no project cannot be
+   * called at all, as it is impossible to determine the project name. Unknown modules are engine
+   * internal constructs that should not use private econstructors at all.
+   */
+  @Test
+  public void privateConstructorCannotBeCalledInUnknownProject() {
+    var src =
+        """
+        type My_Type
+            private Cons data
+        main = My_Type.Cons 42
+        """;
+    try (var ctx = createDefaultContext()) {
+      // TestBase.eval module creates a temporary source - an unknown module without a project.
+      TestBase.evalModule(ctx, src);
+      fail("Should throw a Private_Access panic exception");
+    } catch (PolyglotException e) {
+      assertThat(e.getMessage(), containsString("Private_Access"));
+    }
+  }
+
+  @Test
+  public void privateFieldIsNotExposedToPolyglot() throws IOException {
+    var mainSrc =
+        """
+        type My_Type
+            private Cons data
+        main = My_Type.Cons 42
+        """;
+    var projDir = createProject("My_Project", mainSrc);
+    var mainSrcPath = projDir.resolve("src").resolve("Main.enso");
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, projDir.toAbsolutePath().toString())
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      var mainMod = polyCtx.evalModule(mainSrcPath.toFile());
+      var assocType = mainMod.getAssociatedType();
+      var mainMethod = mainMod.getMethod(assocType, "main").get();
+      var res = mainMethod.execute();
+      assertThat(res.hasMember("data"), is(false));
+    }
+  }
+
+  @Test
+  public void privateConstructorIsNotExposedToPolyglot() throws IOException {
+    var mainSrc = """
+        type My_Type
+            private Cons data
+        """;
+    var projDir = createProject("My_Project", mainSrc);
+    var mainSrcPath = projDir.resolve("src").resolve("Main.enso");
+    try (var ctx =
+        defaultContextBuilder()
+            .option(RuntimeOptions.PROJECT_ROOT, projDir.toAbsolutePath().toString())
+            .build()) {
+      var polyCtx = new PolyglotContext(ctx);
+      var mainMod = polyCtx.evalModule(mainSrcPath.toFile());
+      var myType = mainMod.getType("My_Type");
+      assertThat(myType.hasMember("Cons"), is(false));
```

Comment:
Can you verify that creating `My_Type.Cons 10` in Enso code succeeds, but trying to read `value.readMember("data")` fails then?

Reply:
Extended a test and fixed the interop behavior of `Atom.hasMembers` in eaeb44c25543d1a58bfaa4084e4c8838a1897c04